### PR TITLE
[generate] return past_key_values

### LINKED ISSUE #17016: Optionally return past key values from generate
### Feature request

The idea would be to optionally return `past_key_values` inside the generation objects (`SampleEncoderDecoderOutput`, etc). This could be controlled by a flag called `return_past_key_values` that's passed to `generate` and then forwarded to `sample`, etc.

### Motivation

Perhaps this is niche, but my team and I need often need to obtain the past keys and values when generating in order to manipulate them a bit and then feed them back in for subsequent calls to `generate`. We currently do this with a custom version of `sample`, but this results in us having to copy and paste a lot of code. Would it be possible to allow `past_key_values` to be optionally returned by `generate`?

### Your contribution

If you all approve of the feature idea, I'd be able to implement it and submit a PR.

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

Allows returning `past_key_values` from `generate` when `use_cache=True`.

Like other returned values, `past_key_values` are also returned as `Tuple`, one element per generated token.

Fixes #17016

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -2275,6 +2351,8 @@ def beam_search(
             )
             if model_kwargs["past"] is not None:
                 model_kwargs["past"] = self._reorder_cache(model_kwargs["past"], beam_idx)
+                if return_dict_in_generate:
+                    past_key_values += (model_kwargs["past"],)
```
Comment (patil-suraj) [inline src/transformers/generation_utils.py:2355]: Assumes that if `model_kwargs["past"` is not `None` -> `output_past_key_values == True`

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -2303,6 +2381,8 @@ def beam_search(
             if not output_scores:
                 sequence_outputs["sequence_scores"] = None
 
+            # past_key_values = model_kwargs["past"] if output_past_key_values else None
+
```
Comment (gante) [inline src/transformers/generation_utils.py:2385]:
```suggestion
```

(probably forgotten :) )

**Review by gante — APPROVED** [review]
LGTM 👍

File: tests/generation/test_generation_utils.py [hunk 3]
```diff
@@ -1433,6 +1433,17 @@ def _check_outputs(self, output, input_ids, config, use_cache=False, num_return_
                 use_cache=use_cache,
             )
 
+        # Past Key Value States
+        past_key_values = output.past_key_values
```
Comment (patrickvonplaten) [inline tests/generation/test_generation_utils.py:1437]: Very cool to put it in every test here - good job!

**Comment by patrickvonplaten** [comment]
We'll just need to fix the failing tests now :-) Think you'll have to overwrite this "checking" function in the respective individual test files

**Comment by dblakely** [comment]
Hey there, sorry to nag, but any chance of moving this along?

**Comment by patrickvonplaten** [comment]
If ok for you @gante this would be amazing!

**Comment by shunzh** [comment]
Hi, Thank you all for working on this feature! Is this going to be merged into the main branch soon?

**Comment by gilljon** [comment]
Was this closed because it's now possible to retrieve `past_key_values` or was there another reason?

**Comment by gante** [comment]
@gilljon it is not closed :)

**Comment by gilljon** [comment]
@gante I'm sorry for the confusion! Any idea when it will be merged?

**Comment by sijunhe** [comment]
hi @gante . Any idea when this will be merged? Interested in using it and building something on top of it. I'll happy to put on the finishing touches if needed too!

**Comment by slyalin** [comment]
Hey! Just a friendly reminder. Any chance to get it merged soon?

**Comment by freckletonj** [comment]
I would absolutely **love** this feature! This would open up so much for me, because I have prompts like:

```
prompt = '''
Stuff
* <generate X>
* <generate Y>

Stuff
You said [X], and [Y] previously, now:
* <generate Z>
'''

This is so expensive without `past_key_values`.

So this PR is now Merge-Conflicting, and I tried applying the patch but upon inspection, it's quite severely out of date now.

**Is there another way to accomplish this?**

I notice that `model.forward` typically allows to return `past_key_values`. But then I... have to make use of a sampling alg myself? Would this be the best way without needing upstream changes, and if so, how can I chain together `model.forward` and a sampler?

**Comment by freckletonj** [comment]
Is this ticket dead because some other technique exists already for returning and reusing `past_key_values`? This is a killer feature.

**Comment by freckletonj** [comment]
The following PR is more up to date: https://github.com/huggingface/transformers/pull/25086

**Comment by gante** [comment]
(deprecated in favor of #25086)

**Comment by gante** [comment]
Hey folks 👋 

#25086 was merged.

If you install from `main` and add `return_dict_in_generate=True` to `generate`, `past_key_values` will be part of the output, assuming your model is configured with `use_cache=True` (the default).

You can then pass `past_key_values` to `generate` to continue generating!

**Comment by nevakrien** [comment]
I cant get it to work with intel neural_chat what vartion was this on?

**Pull request closed without merging**